pkg/fs/fs.go

```diff
@@ -39,37 +39,35 @@ type Filesystem struct {
 	minFreeDisk        int64
 	rwLock             *sync.RWMutex
 	multiparts         *Multiparts
-	listObjectMap      map[ListObjectParams][]ObjectInfoChannel
+	listObjectMap      map[ListObjectParams][]*treeWalker
 	listObjectMapMutex *sync.Mutex
 }
 
-func (fs *Filesystem) pushListObjectCh(params ListObjectParams, ch ObjectInfoChannel) {
+func (fs *Filesystem) pushTreeWalker(params ListObjectParams, walker *treeWalker) {
```

Comment:
`ObjectInfoChannel` is misleading as it is not a channel but a structure.

changed the variable names so that it's more readable:
* treeWalk() function is the one that recurses(walks) into the directory tree.
* treeWalker is the go-routine that does treeWalk and is associated with structure treeWalker{}
* treeWalker go-routine is started by startTreeWalker()
* treeWalker go-routine pushes treeWalkResult{} up the channel 
* treeWalkResult{} contains the desired ObjectInfo structure and also other info like `err`, `end`.

the entire logic also takes care of prefix and marker.

this problem can be expressed as a classic case of recursion.